server/src/main/java/org/candlepin/model/ProductCurator.java

```diff
@@ -228,20 +232,20 @@ public Product getByName(Owner owner, String name) {
      */
     public int hydratePoolProvidedProducts(Iterable<Pool> pools) {
         int count = 0;
-
         if (pools != null && pools.iterator().hasNext()) {
             EntityManager entityManager = this.getEntityManager();
-            Cache<String, Product> productCache = this.candlepinCache.getProductCache();
 
             // We use these set for both product UUIDs and product instances, once populated
-            Map<String, Set> poolProvidedProducts = new HashMap<>();
-            Map<String, Set> poolDerivedProvidedProducts = new HashMap<>();
+            Map<String, Set<String>> poolProvidedProducts = new HashMap<>();
+            Map<String, Set<String>> poolDerivedProvidedProducts = new HashMap<>();
+            Map<String, Product> allProducts = new HashMap<>();
 
             Set<String> poolIds = new HashSet<>();
-            Set<String> productUuids = new HashSet<>();
 
             for (Pool pool : pools) {
                 poolIds.add(pool.getId());
+                poolProvidedProducts.put(pool.getId(), new HashSet<>());
```

Comment:
If we're pre-populating the pool->product maps, then the poolIds set is unnecessary, since we can just use the keyset of either map in its place.

Reply:
Only this remains. Other changes look good.